File: utils/event_engine.py
class EventEngine:
    """Load and trigger events from a YAML or JSON configuration."""

    def __init__(self, path: str | None = None) -> None:
        """Initialize the engine with an optional file path."""
        self.events: list[dict] = self._load(path) if path else []
        self.by_step: dict[int, list[dict]] = {}
        for evt in self.events:
            step = int(evt.get("step", 0))
            self.by_step.setdefault(step, []).append(evt)

    def _load(self, path: str) -> list[dict]:
        import json
        if path.endswith((".yaml", ".yml")):
            try:
                import yaml  # type: ignore
            except Exception as e:  # pragma: no cover - optional dep
                raise ImportError("PyYAML is required for YAML events") from e
            with open(path) as f:
                data = yaml.safe_load(f) or []
        else:
            with open(path) as f:
                data = json.load(f)
        if isinstance(data, dict):
            data = data.get("events", [])
        if not isinstance(data, list):
            raise ValueError("Event file must contain a list")
        return data

    def trigger_events(self, step: int, sim) -> None:
        """Execute any events scheduled for ``step``."""
        events = self.by_step.pop(step, [])
        for evt in events:
            etype = evt.get("type")
            if etype == "market_shock":
                severity = evt.get("severity")
                sim.trigger_market_shock(severity)
            elif etype == "marketing_campaign":
                self._run_campaign(evt, sim)
            elif etype == "create_proposal":
                self._create_proposal(evt, sim)

    def add_event(self, event: dict) -> None:
        """Schedule a new event at runtime."""
        self.events.append(event)
        step = int(event.get("step", 0))
        self.by_step.setdefault(step, []).append(event)

    def list_events(self) -> list:
        """Return all loaded events."""
        return list(self.events)

    def _run_campaign(self, evt: dict, sim) -> None:
        from data_structures.marketing_events import (
            DemandBoostCampaign,
            RecruitmentCampaign,
            SocialMediaCampaign,
            ReferralBonusCampaign,
        )

        ctype = evt.get("campaign_type", "social_media")
        budget = float(evt.get("budget", 50))
        if ctype == "demand_boost":
            camp = DemandBoostCampaign(
                sim.dao,
                budget,
                evt.get("price_boost", 0.1),
            )
        elif ctype == "recruitment":
            camp = RecruitmentCampaign(
                sim.dao,
                budget,
                int(evt.get("recruits", 2)),
            )
        elif ctype == "referral_bonus":
            camp = ReferralBonusCampaign(
                sim.dao,
                budget,
                int(evt.get("recruits", 1)),
                float(evt.get("bonus", 10)),
            )
        else:
            camp = SocialMediaCampaign(
                sim.dao,
                budget,
                evt.get("price_boost", 0.05),
            )
        camp.execute(sim)

    def _create_proposal(self, evt: dict, sim) -> None:
        from utils.proposal_utils import create_random_proposal
        creator = sim.dao.members[0] if sim.dao.members else None
        title = evt.get("title", "Event Proposal")
        proposal = create_random_proposal(sim.dao, creator, title_prefix=title)
        if proposal:
            sim.dao.add_proposal(proposal)
            if creator:
                creator.submit_proposal(proposal)

File: lib/utils/event-engine.ts
// Event Engine - load and trigger events from configuration
// Port from utils/event_engine.py

import type { DAOModel } from '../engine/model';
import { createRandomProposal } from './proposal-utils';

export interface EventConfig {
  step: number;
  type: string;
  [key: string]: any;
}

export class EventEngine {
  events: EventConfig[] = [];
  byStep: Map<number, EventConfig[]> = new Map();

  constructor(events?: EventConfig[]) {
    if (events) {
      this.loadEvents(events);
    }
  }

  /**
   * Load events from configuration array
   */
  loadEvents(events: EventConfig[]): void {
    this.events = events;
    this.byStep.clear();

    for (const evt of events) {
      const step = evt.step || 0;
      if (!this.byStep.has(step)) {
        this.byStep.set(step, []);
      }
      this.byStep.get(step)!.push(evt);
    }
  }

  /**
   * Load events from JSON string
   */
  loadFromJson(json: string): void {
    const data = JSON.parse(json);
    let events: EventConfig[];

    if (Array.isArray(data)) {
      events = data;
    } else if (data.events && Array.isArray(data.events)) {
      events = data.events;
    } else {
      throw new Error('Event data must be an array or contain an "events" array');
    }

    this.loadEvents(events);
  }

  /**
   * Trigger all events scheduled for the given step
   */
  triggerEvents(step: number, sim: DAOModel): void {
    const events = this.byStep.get(step);
    if (!events) {
      return;
    }

    // Remove from map after retrieving
    this.byStep.delete(step);

    for (const evt of events) {
      const eventType = evt.type;

      switch (eventType) {
        case 'market_shock':
          this.triggerMarketShock(evt, sim);
          break;

        case 'create_proposal':
          this.createProposal(evt, sim);
          break;

        default:
          console.warn(`Unknown event type: ${eventType}`);
      }
    }
  }

  /**
   * Trigger a market shock event
   */
  private triggerMarketShock(evt: EventConfig, sim: DAOModel): void {
    const severity = evt.severity || 0;

    if (typeof (sim as any).triggerMarketShock === 'function') {
      (sim as any).triggerMarketShock(severity);
    } else {
      // Fallback: manually adjust price
      const oldPrice = sim.dao.treasury.getTokenPrice('DAO_TOKEN');
      const newPrice = oldPrice * (1 + severity);
      sim.dao.treasury.updateTokenPrice('DAO_TOKEN', newPrice);

      if (sim.eventBus) {
        sim.eventBus.publish('market_shock', {
          step: sim.currentStep,
          severity,
          oldPrice,
          newPrice,
        });
      }
    }
  }

  /**
   * Create a proposal from event configuration
   */
  private createProposal(evt: EventConfig, sim: DAOModel): void {
    const creator = sim.dao.members.length > 0 ? sim.dao.members[0] : undefined;
    const title = evt.title || 'Event Proposal';

    const proposal = createRandomProposal(sim.dao, creator, title);

    if (proposal) {
      sim.dao.addProposal(proposal);
    }
  }

  /**
   * Add a new event at runtime
   */
  addEvent(event: EventConfig): void {
    this.events.push(event);

    const step = event.step || 0;
    if (!this.byStep.has(step)) {
      this.byStep.set(step, []);
    }
    this.byStep.get(step)!.push(event);
  }

  /**
   * Get all loaded events
   */
  listEvents(): EventConfig[] {
    return [...this.events];
  }

  /**
   * Get events scheduled for a specific step
   */
  getEventsForStep(step: number): EventConfig[] {
    return this.byStep.get(step) || [];
  }

  /**
   * Clear all events
   */
  clear(): void {
    this.events = [];
    this.byStep.clear();
  }

  /**
   * Export events as JSON
   */
  toJson(): string {
    return JSON.stringify({ events: this.events }, null, 2);
  }
}
